rfc/0009-audit-checks.md

```diff
@@ -137,82 +120,52 @@ spec:
+write at the same time. This will cause an error in the latest write. A retry mechanism should be implemented in the
+Background Audit Scanner to handle this scenario.
 
-The reports should contains some minimum information listed below:
+No real time audit info. In future iterations Audit Scanner could be watching all resources and updating reports in real time.
 
- - Policy name/ID
- - Policy server name/ID
- - Policy validation result (e.g. `success`, `fail`, etc)
- - Policy failure error message, if it exists. The same error message shown to the users if they try to apply an invalid resource in the cluster.
- - Resource identification
- - Policy mode ( `monitor` or `protect` mode)
- - Timestamp of the validation
+Many more calls to Policy Server and the Kubernetes api server. Caching should help, but still many more calls will be made.
 
-Any additional information is optional.
+`PolicyReports` size could be big if there are many resources and policies in the same namespace. More info on how big it 
+could be [here](https://github.com/kubewarden/rfc/pull/13#issuecomment-1253744208)
 
-# Drawbacks
+# Alternatives
 
-It's not possible to define if a policy with `audit` field `true` is really
-validating the resources in the new added `resources` field before running the audit check.
+## Background Auditor Scanner is only service that updates PolicyReports
 
-Download policies one more time. Once for the policy server running in audit mode
-and others for the other policy servers
+PolicyServer would not store evaluations in PolicyReport. Instead, it would call the Auditor Scanner using a rest API.
+This means Background Audit Scanner should always be running, and it can't be a `CronJob`.
 
-# Alternatives
+If auditor is not available report data is lost. This data will be recovered in the next background scan.
 
-Additional to the `PolicyReport` report, policy server could add an annotation to the
-resources violating policies. However, this will require additional permissions to the policy server.
+Background Auditor Scanner is the only service writing `PolicyReports`, which makes easier to handle concurrent writes.
 
-Policies could have a dedicated function to evaluate resources in the audit checks
+## Run background scan in PolicyServer
 
-If the number of resources for an individual policy is too high, the resources could
-be batched and multiple instances of the same policy could be run at the same time
-temporally to speed up the policy validations.  This horizontal scaling for all the
-policies under demand is a good topic for another RFC or idea discussion for the future.
+No need to create Background Auditor Scanner. Everything is done in `PolicyServer`. PolicyServer will take care 
+of doing the background scan.
 
+If background scan takes too long `PolicyServer` might be slower for evaluation requests, which would slow down creation 
```

Comment:
Follow-up:
Kubernetes dry-run requests also help in evaluating requests correctly when they are mutated, as dry-run requests will get mutated by the mutating policy, and then validated by the rest of policies, closely simulating reality. If not using dry-run requests, policies would evaluate non-mutated requests when they would have been mutated.

Reply:
> I don't understand what you mean here. Can you explain it, please?

My musing is about, how does a PolicyServer know which Policy they need to use to evaluate a request? This was done previously by the webhook registration and Kube API server, we need to re-implement this. Moreover, how does it work for mutated requests? do we re-issue the mutated audit request for validating policies, as Kube API server does?

Reply:
We know which policy-server and which resources a policy will evaluate just by looking at the policy spec. We know the endpoint for each policy in the relevant policy-server because they always follow the same naming convention.
Mutation won't happen when running the background audit check. Background Auditor Check requests evaluation for a specific resource to a PolicyServer. If the response contains a mutation it will be ignored.

See [this](https://github.com/kubewarden/rfc/pull/10#discussion_r982073919) for dry-run

Reply:
> We know which policy-server and which resources a policy will evaluate just by looking at the policy spec. We know the endpoint for each policy in the relevant policy-server because they always follow the same naming convention.

I suppose we need to look for namespaced `AdmissionPolicies`,  honour `spec.namespaceSelector`, `spec.objectSelector`, `spec.matchPolicy == equivalent` too. We ought to document this on the list of audited policies.

> Mutation won't happen when running the background audit check.

We ought to document this in the RFC, as part of the list of non-audited requests.

Reply:
added in https://github.com/kubewarden/rfc/pull/10/commits/3172607122dadbd95a99e3433219816d2f3f9c59 
please let me know if you think there is still something missing

Reply:
Thanks! what about `spec.objectSelector`, `spec.matchPolicy == equivalent`? Do we need to do something specific?
What if someone is using `objectSelector` to target objects with a `prod` label? We would running more policy evaluations than needed, and there would be results that shouldn't be there.